Allow external directories for 'linuxkit pkg build'

rn opened the pull request:
This adds adds a new `sources` field to the `build.yml` where one can specify additional source directories for the package build outside the package directory. This is useful for sharing files, such as vendored go code, between packages.

If `sources` are specified, the package hash is calculated as the `sha1` of the strings of the git tree hashes of all the externals sources and the package directory. Therefore, `sources` need to be under git revision control.

`sources` are specified as a `src:dst` pair, where `src` can be a relative path the package directory and `dst` is the destination with the docker build context.

The `sources` features is implemented by `tar`ing up (and rewriting the paths) of the sources and pipe them to `stdin` of the `docker` process used to build the package.

There should be no changes for packages not using the `sources` features.

resolves #3108

![whales](https://user-images.githubusercontent.com/3338098/43212788-dba7ebbc-902c-11e8-9be4-902a33fb9bab.jpg)

rn commented:
@justincormack thanks for the review. I'll wait till tomorrow for @ijc eyes. This is mostly his code I modified

ijc commented on docs/packages.md:
```diff
@@ -37,6 +37,7 @@ A package source consists of a directory containing at least two files:
 - `image` _(string)_: *(mandatory)* The name of the image to build
 - `org` _(string)_: The hub/registry organisation to which this package belongs
 - `arches` _(list of string)_: The architectures which this package should be built for (valid entries are `GOARCH` names)
+- `sources` _(list of strings)_: Optionally specifies sources for the package build which are outside the package directory. The format is `src:dst`, where `src` can be a relative path to the package directory and `dst` is the destination in the build context. This is useful for sharing files, such as vendored go code, between packages.
```
I think `extra-sources` would be a more accurate name.

> where `src` can be a relative path to the package directory

"can" or "must"? I would also word it as "path relative to the".

ijc commented on src/cmd/linuxkit/pkglib/build.go:
```diff
@@ -191,3 +201,62 @@ func (p Pkg) Build(bos ...BuildOpt) error {
+			}
+			defer f.Close()
+
+			h, err := tar.FileInfoHeader(i, "")
```
In `linuxkit build` we force `PAX` headers, IIRC for reasons to do with path length limitations. Should we do the same here?

rn replied:
turns out `dockerd` doesn't like it:
```
Sending build context to Docker daemon   16.9kB
Error response from daemon: Dockerfile parse error line 1: unknown instruction: /PAXHEADERS.00000000000000000000000000000013200000000000010217
exit status 1
```

ijc replied:
Interesting!

I guess the tarfile is all relative to the root of the build context, so it won't matter if `cwd` is super long and we only worry about very deep paths within the actual context itself (which is less likely).

ijc commented on src/cmd/linuxkit/pkglib/build.go:
```diff
@@ -191,3 +201,62 @@ func (p Pkg) Build(bos ...BuildOpt) error {
+		f := func(p string, i os.FileInfo, err error) error {
+			if err != nil {
+				log.Warnf("ctx: Walk error on %s: %v", p, err)
+				return nil
```
This function appears to warn and eat all the errors -- should it not return the error and fail the build?

rn replied:
There does not seem to be a way to return an error all the way back to the pipe and no way yo handle the error. I could change this to log.Fatal()` or similar to exit

ijc replied:
Returning an error in this function propagates to the error return of the `filepath.Walk`, you just need to handle it there instead of ignoring it. I think.

Maybe you need to propagate it further, like back the `if dr.ctx != nil {` block goroutine, an `errgroup` there would save you messing around with channels manually.

ijc commented on src/cmd/linuxkit/pkglib/pkglib.go:
```diff
@@ -169,6 +172,43 @@ func NewFromCLI(fs *flag.FlagSet, args ...string) (Pkg, error) {
+			return Pkg{}, fmt.Errorf("Bad source format in %s", source)
+		}
+
+		// Make sure the srcPath
```
missing the end of the sentence?

ijc commented on src/cmd/linuxkit/pkglib/pkglib.go:
```diff
@@ -169,6 +172,43 @@ func NewFromCLI(fs *flag.FlagSet, args ...string) (Pkg, error) {
+	if len(pi.Sources) != 0 {
+		// If we have external sources ensure the package path is included first
+		sources = []string{pkgPath + ":/"}
+	}
```
If we made it so that sources always contains at least the pkgPath (and nothing else if `pi.Sources` is unused) then wouldn't that remove a bunch of conditionality elsewhere -- e.g. the building bit would then always pass the context on stdin even in the single source case.

rn replied:
Almost certainly. I contemplated it but initially wanted to add this feature but leave the other code path untouched, but maybe now it's better to always use the build context

ijc replied:
I think it is necessary to leave the behaviour the same, but changing the internals as needed to make the new stuff fit cleanly is the right thing to do.

ijc commented on src/cmd/linuxkit/pkglib/pkglib.go:
```diff
@@ -169,6 +172,43 @@ func NewFromCLI(fs *flag.FlagSet, args ...string) (Pkg, error) {
+		// Make sure the srcPath
+		srcPath := tmp[0]
+		if !filepath.IsAbs(srcPath) {
+			srcPath, err = filepath.Abs(filepath.Join(pkgPath, srcPath))
```
`filepath.Join(pkgPath, srcPath)` is already absolute (and `.Clean`), so `filepath.Abs` here does nothing.

ijc commented on src/cmd/linuxkit/pkglib/build.go:
```diff
@@ -191,3 +201,62 @@ func (p Pkg) Build(bos ...BuildOpt) error {
+			}
+
+			if i.IsDir() {
+				return nil
```
I guess the untar on the other end automatically creates missing dirs, but aren't you throwing away ownership and perms etc? Is there any harm in outputting the directories too?

I'm not entirely sure that the remaining code handles symlinks and the like ok, it seems to either collapse them to files or output a symlink header followed by some actual data, not sure what the result is.

rn replied:
Let me try to refactor this a little bit. I got an error when hitting directories, but I think I now know a better way to handle this

rn commented:
@ijc thanks for the review. I think I have addressed all the comments:
- `sources` -> `extra-sources`
- Always use the build context. Makes the code a little simpler
- Propagate errors from the build context `Copy()` function and then use an `errgroup`
- plus the minor bits and pieces

ijc commented on docs/packages.md:
```diff
@@ -37,6 +37,7 @@ A package source consists of a directory containing at least two files:
 - `image` _(string)_: *(mandatory)* The name of the image to build
 - `org` _(string)_: The hub/registry organisation to which this package belongs
 - `arches` _(list of string)_: The architectures which this package should be built for (valid entries are `GOARCH` names)
+- `extra-sources` _(list of strings)_: Additional sources for the package outside the package directory. The format is `src:dst`, where `src` can be a relative to the package directory and `dst` is the destination in the build context. This is useful for sharing files, such as vendored go code, between packages.
```
"can be a relative" → "can be relative".

Is there really a usecase for absolute paths here?

rn replied:
it doesn't matter. In most cases it will be relative but it is not enforced to be an absolute path.

ijc commented on src/cmd/linuxkit/pkglib/build.go:
```diff
@@ -191,3 +199,62 @@ func (p Pkg) Build(bos ...BuildOpt) error {
+			if err != nil {
+				return fmt.Errorf("ctx: Converting FileInfo for %s: %v", p, err)
+			}
+			h.Name = path.Join(dst, strings.TrimPrefix(p, src))
```
`p` at this point has, I think, been [`Clean`ed](https://golang.org/pkg/path/filepath/#Clean) relative to the original `src`.

If `src` was unclean, perhaps having `//` or "/../" in it or similar then `TrimPrefix` won't DTRT. It looks like [`filepath.Rel`](https://golang.org/pkg/path/filepath/#Rel) might do exactly what you want. It might be sufficient to `filepath.Clean` both `src` and `dst` yourself, but I think `filepath.Rel` is probably better.

ijc commented on src/cmd/linuxkit/pkglib/pkglib.go:
```diff
@@ -169,6 +172,36 @@ func NewFromCLI(fs *flag.FlagSet, args ...string) (Pkg, error) {
+	sources := []string{pkgPath + ":/"}
+
+	for _, source := range pi.ExtraSources {
+		tmp := strings.Split(source, ":")
```
You do this splitting and what have you in two places, you could consider making sources be a `[]struct Foo {src, dst string}` perhaps and accumulating it once here.

ijc commented on src/cmd/linuxkit/pkglib/pkglib.go:
```diff
@@ -169,6 +178,36 @@ func NewFromCLI(fs *flag.FlagSet, args ...string) (Pkg, error) {
+			return Pkg{}, fmt.Errorf("Bad source format in %s", source)
+		}
+
+		srcPath := tmp[0]
```
An absolute `srcPath` is never cleaned AFAICT (a relative one gets it implicitly in `filepath.Join`).

ijc commented on src/cmd/linuxkit/pkglib/pkglib.go:
```diff
@@ -169,6 +178,36 @@ func NewFromCLI(fs *flag.FlagSet, args ...string) (Pkg, error) {
 	})
 
+	var srcHashes string
+	sources := []pkgSource{pkgSource{src: pkgPath, dst: "/"}}
```
I think `gofmt` is going to want to make this  `[]pkgSource{{src: pkgPath, dst: "/"}}` which is more concise (the inner type is implicit)

ijc commented on src/cmd/linuxkit/pkglib/pkglib.go:
```diff
@@ -169,6 +178,36 @@ func NewFromCLI(fs *flag.FlagSet, args ...string) (Pkg, error) {
+		}
+
+		srcHashes += h
+		sources = append(sources, pkgSource{src: srcPath, dst: tmp[1]})
```
`dst` is also never `Clean`d (which may or may not matter).

rn commented:
Now cleaning paths and fix gofmt issue
